craft_application/services/service_factory.py

```diff
@@ -89,6 +89,15 @@ def __getattr__(self, service: str) -> services.AppService:
                 raise ValueError(
                     f"{cls.__name__} requires a project to be available before creation."
                 )
+            for field in self.app.mandatory_adoptable_fields:
+                if (
+                    not getattr(self.project, field, None)
```

Comment:
this feels awkward, why is this check in the factory?

Reply:
Maybe because it's not easy to put into pydantic?

Reply:
But then I think this check would be better placed somewhere like Application.get_project(), I don't think the factory should be scrutinizing the project like this

Reply:
My suggestion is [here](https://github.com/canonical/craft-application/blob/main/craft_application/application.py#L280), after the Project is created but before it's set on `self.__project` (so a step before the project is set on the factory)